clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3999,7 +4073,8 @@ void Sema::CheckSYCLKernelCall(FunctionDecl *KernelFunc,
   if (KernelObj->isInvalidDecl())
     return;
 
-  bool IsSIMDKernel = isESIMDKernelType(KernelObj);
+  bool IsSIMDKernel =
+      isESIMDKernelType(KernelCallOperator, KernelObj, KernelFunc, *this);
```

Comment:
From my understanding, this code is not capable of understanding whether kernel is ESIMD kernel if it is a functor. How this will be handled?

Reply:
`CheckSYCLKernelCall` is called very early, before templates are instantiated, so the function with `sycl_kernel` attribute doesn't have a body yet. I suppose we can't find `operator()` from here using a visitor.
At some point this might be worse, but only in CheckSYCLKernelCall function. 
In other places the new approach is better since it finds the right operator, not the first found overload.

Reply:
@Fznamznon For Explicit SIMD kernels, in the old code, we only looked at the kernel function object type and returned the first found function call operator, `operator()`. ( for kernels defined as lambdas as well as functors).
It dint' matter when we reached `isESIMDKernelType` via `CheckSYCLKernelCall` because we did not have the need to check the function body of the sycl kernel function ( the one with `sycl_kernel` attribute).

However, in the new approach, since we check the function body of the sycl kernel function to identify which version of operator()() call is invoked in user code, we definitely need the sycl kernel function to be fully defined/instatiated.

The problem we have here is though, the sycl kernel function is still a un-instantiated template in 
`CheckSYCLKernelCall` and so there is no way we can find the right call operator function from the sycl kernel function found here.

Currently, with our new approach, we get  `IsSIMDKernel = false` in `CheckSYCLKernelCall` for functors with multiple SIMD call operators.

This is incorrect and will cause problems with diagnostics handled in `SyclKernelFieldChecker` and `SyclKernelArgsSizeChecker`

I also checked with ESIMD folks and made sure that it is indeed legal code to have kernel functors with multiple SIMD call operators defined.

The old code was incorrect for ESIMD kernels as well with regards to kernel functors with multiple SIMD call operators defined.

I think we can try setting the IsSIMD bool in `SyclKernelFieldChecker` and `SyclKernelArgsSizeChecker` from `ConstructOpenCLKernel`

Please let me know if you have other ideas.